src/Controls/src/Core/HybridWebView/HybridWebView.cs

```diff
@@ -15,40 +25,215 @@ public class HybridWebView : View, IHybridWebView
+			}
+
+			var messageType = rawMessage.Substring(0, indexOfPipe);
+			var messageContent = rawMessage.Substring(indexOfPipe + 1);
+
+			switch (messageType)
+			{
+				case "InvokeMethodCompleted":
+					{
+						var sections = messageContent.Split('|');
+						var taskId = sections[0];
+						var result = sections[1];
+						AsyncTaskCompleted(taskId, result);
+					}
+					break;
+				case "RawMessage":
+					RawMessageReceived?.Invoke(this, new HybridWebViewRawMessageReceivedEventArgs(messageContent));
+					break;
+				default:
+					throw new ArgumentException($"The message type '{messageType}' is not recognized.", nameof(rawMessage));
+			}
 		}
 
 		/// <summary>
 		/// Raised when a raw message is received from the web view. Raw messages are strings that have no additional processing.
 		/// </summary>
 		public event EventHandler<HybridWebViewRawMessageReceivedEventArgs>? RawMessageReceived;
 
+		/// <summary>
+		/// Sends a raw message to the code running in the web view. Raw messages have no additional processing.
+		/// </summary>
+		/// <param name="rawMessage"></param>
 		public void SendRawMessage(string rawMessage)
 		{
-			Handler?.Invoke(nameof(IHybridWebView.SendRawMessage), rawMessage);
+			Handler?.Invoke(
+				nameof(IHybridWebView.SendRawMessage),
+				new HybridWebViewRawMessage
+				{
+					Message = rawMessage,
+				});
+		}
+
+		private int _invokeTaskId;
+		private Dictionary<string, TaskCompletionSource<string>> asyncTaskCallbacks = new Dictionary<string, TaskCompletionSource<string>>();
+
+		/// <summary>
+		/// Handler for when the an Async JavaScript task has completed and needs to notify .NET.
+		/// </summary>
+		private void AsyncTaskCompleted(string taskId, string result)
+		{
+			//Look for the callback in the list of pending callbacks.
+			if (!string.IsNullOrEmpty(taskId) && asyncTaskCallbacks.ContainsKey(taskId))
+			{
+				//Get the callback and remove it from the list.
+				var callback = asyncTaskCallbacks[taskId];
+				callback.SetResult(result);
+
+				//Remove the callback.
+				asyncTaskCallbacks.Remove(taskId);
+			}
+		}
+
+		/// <summary>
+		/// Invokes a JavaScript method named <paramref name="methodName"/> and optionally passes in the parameter values
+		/// specified by <paramref name="paramValues"/>.
+		/// </summary>
+		/// <param name="methodName">The name of the JavaScript method to invoke.</param>
+		/// <param name="paramValues">Optional array of objects to be passed to the JavaScript method.</param>
+		/// <param name="paramJsonTypeInfos">Optional array of metadata about serializing the types of the parameters specified by <paramref name="paramValues"/>.</param>
+		/// <returns>A string containing the return value of the called method.</returns>
+#if WINDOWS || ANDROID || IOS || MACCATALYST
+		public async Task<string?> InvokeJavaScriptAsync(string methodName, object?[]? paramValues, JsonTypeInfo?[]? paramJsonTypeInfos = null)
```

Comment:
There should be almost no `#if` things in Controls as it really is there for cases where it is not possible to do it any way else. 

No matter if the operation is not supported, the Controls layer should always send the command to the handler where the work is done. If the handler decides it is not supported, then the handler can throw. This also allows the user to override the full operation if need be to make it work.